Add support for AssumeRoleWithWebIdentity

partcyborg opened the pull request:
<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Add support for STS [AssumeRoleWithWebIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html).

Includes new config option `iam_web_identity_token` which takes either a WebIdentity token (designed to be passed in with `get_env()`), or a path to a file containing a WebIdentity token.

Fixes #2585.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added attribute `iam_web_identity_token` to support AssumeRoleWithWebIdentity 

### Migration Guide

<!-- Important: If you made any backward incompatible changes, then you must write a migration guide! -->

partcyborg commented:
FYI: I just sent out a PR that implements this.

partcyborg commented:
Sorry, I have no idea how this got closed. I certainly don't remember doing it.

denis256 commented:
```
aws_helper/config.go:5:2: SA1019: "io/ioutil" has been deprecated since Go 1.19: As of Go 1.16, the same functionality is now provided by package [io] or package [os], and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
        "io/ioutil"
        ^
```

partcyborg commented:
> ```
> aws_helper/config.go:5:2: SA1019: "io/ioutil" has been deprecated since Go 1.19: As of Go 1.16, the same functionality is now provided by package [io] or package [os], and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
>         "io/ioutil"
>         ^
> ```
Thanks for the review! I have replaced calls to `ioutil.ReadFile` with calls to `os.ReadFile`.

partcyborg commented:
@denis256 any chance I could get a followup review? I addressed the comments you left previously.  Thanks in advance!

partcyborg commented:
Hi @denis256 thanks for picking this up again.

I have resolved the merge conflicts (and updated the new flag name per the new naming convention).

Please take another look.

partcyborg commented:
Synced this branch against master again, resolving a small merge conflict.

Hey @denis256 I would really appreciate a followup review so I don't have to keep checking syncing this to avoid merge conflicts. Please and thank you!

denis256 commented on aws_helper/config.go:
```diff
@@ -188,18 +231,47 @@ func AssumeIamRole(iamRoleOpts options.IAMRoleOptions) (*sts.Credentials, error)
+			return nil, errors.WithStackTrace(err)
+		}
+		return resp.Credentials, nil
+	} else {
```
I think else is not required since `return` is used

partcyborg replied:
Thanks for the review. I have removed the unnecessary else clause

denis256 commented:
It is possible to implement integration tests to track that this functionality will work over time?

partcyborg commented:
> It is possible to implement integration tests to track that this functionality will work over time?

Yes, depending on your CI environment.  For example, we will be using this functionality to replace shell script calls to `aws sts assume-role-with-web-identity` in our CI/CD environment.

Do you have integration tests for the current `iam_role` functionality? I would happily modify them for this, but I don't see anything in the repo that uses it.


In order to do this, you need to
- Setup an AWS IAM OpenID Connect provider in the account
- Setup an IAM role for your terragrunt to execute under
- Grant AssumeRole permission to the OpenID Connect provider using an AssumeRolePolicyDocument entry that looks like this 
```
{
    "Effect": "Allow",
    "Principal": {
        "Federated": "arn:aws:iam::<account>:oidc-provider/<name>"
    },
    "Action": "sts:AssumeRoleWithWebIdentity",
    "Condition": {
        "StringLike": {
            "<OIDC provider>:sub": "<repo path from your provider documentation>"
        }
    }
}
```

Then point `iam_web_identity_token` at the either environment variable or filepath provided by the provider.

Here are the docs for setting this up in gitlab: https://docs.gitlab.com/ee/ci/cloud_services/aws/

Here are the docs for setting this up in github: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services

denis256 commented:
Hello,
then this PR should be extended with integration tests which will use `AssumeRoleWithWebIdentity`, required IAM role can be read from environment variables, similar to  `TestTerragruntAssumeRoleDuration`

yhakbar commented:
To be clear, @partcyborg , we don't need these tests to be running in our CircleCI configuration for this pull request.

This integration test only has to work in the context of your CI system. Run it there, show the configurations you used (masking any account identifiers, etc), and share the test output with us so that we can consider this functionality tested.

They should require the minimum set of dependencies like an environment variable and an IAM role. It should also be accompanied with example CI setups in the docs that we can share with the community.

Once we have these, we can independently verify the tests, then merge this PR. In a subsequent PR, we can setup the integration tests into the CI configurations for this repository.

partcyborg commented:
> To be clear, @partcyborg , we don't need these tests to be running in our CircleCI configuration for this pull request.
> 
> This integration test only has to work in the context of your CI system. Run it there, show the configurations you used (masking any account identifiers, etc), and share the test output with us so that we can consider this functionality tested.
> 
> They should require the minimum set of dependencies like an environment variable and an IAM role. It should also be accompanied with example CI setups in the docs that we can share with the community.
> 
> Once we have these, we can independently verify the tests, then merge this PR. In a subsequent PR, we can setup the integration tests into the CI configurations for this repository.

Thanks for the more detailed explanation @yhakbar

Just to be sure I understand, you want me to
- Create a new integration test (in the `test` directory) that makes use of `iam_role` with `iam_web_identity_token`
- Setup a CI pipeline in our CI infrastructure that runs this test

Is this correct?

yhakbar commented:
> > To be clear, @partcyborg , we don't need these tests to be running in our CircleCI configuration for this pull request.
> > This integration test only has to work in the context of your CI system. Run it there, show the configurations you used (masking any account identifiers, etc), and share the test output with us so that we can consider this functionality tested.
> > They should require the minimum set of dependencies like an environment variable and an IAM role. It should also be accompanied with example CI setups in the docs that we can share with the community.
> > Once we have these, we can independently verify the tests, then merge this PR. In a subsequent PR, we can setup the integration tests into the CI configurations for this repository.
> 
> Thanks for the more detailed explanation @yhakbar
> 
> Just to be sure I understand, you want me to
> 
>     * Create a new integration test (in the `test` directory) that makes use of `iam_role` with `iam_web_identity_token`
> 
>     * Setup a CI pipeline in our CI infrastructure that runs this test
> 
> 
> Is this correct?

That's basically it! If you could also provide some documentation updates to help other customers set up their usage of the feature, with example workflows, etc that would be great.

partcyborg commented:
> To be clear, @partcyborg , we don't need these tests to be running in our CircleCI configuration for this pull request.
> 
> This integration test only has to work in the context of your CI system. Run it there, show the configurations you used (masking any account identifiers, etc), and share the test output with us so that we can consider this functionality tested.
> 
> They should require the minimum set of dependencies like an environment variable and an IAM role. It should also be accompanied with example CI setups in the docs that we can share with the community.
> 
> Once we have these, we can independently verify the tests, then merge this PR. In a subsequent PR, we can setup the integration tests into the CI configurations for this repository.

Thanks @yhakbar for the detailed explanation.

Just so that I am clear what is being asked for here, you want me to:
- Write an integration test, similar to `TestTerragruntAssumeRoleDuration` that makes use of `iam_role` and `iam_web_identity_token`
- Configure a CI pipeline in our environment (we use Gitlab) that runs this test to ensure that it works correctly.
- Include documentation for how to set this up in common git providers
- 

Is this correct?

partcyborg commented:
Hello @yhakbar and @denis256 

I have added two integration tests for `iam_web_identity_token`:
- `TestTerragruntAssumeRoleWebIdentityEnv` tests the functionality when the token is provided as input from an environment variable
- `TestTerragruntAssumeRoleWebIdentityFile` tests the functionality when the token is provided as a local file on disk

I also updated the documentation to include instructions for how to set this up in GitHub, GitLab, and CircleCI.

To run the tests, I uploaded my fork of the terragrunt repo to my company's private gitlab project. I was then able to successfully execute both integration tests using an IAM role configured for web identity using the documentation referenced in my changes to the terragrunt documentation in this pull request.

I have included the GitLab pipeline config config used to run both tests, along with the pipeline's output (showing that the tests passed) in the following gist: https://gist.github.com/partcyborg/370d66dfea927052e27fa00f2e378019

The pipeline's raw output is also accessible [here](https://cdn.artifacts.gitlab-static.net/71/ad/71ade460295c5db115e89072f30dd918f37478c90dd9dfcf62e41266952ae21c/2024_06_05/7020555507/7663851467/job.log?response-content-type=text%2Fplain%3B%20charset%3Dutf-8&response-content-disposition=inline&Expires=1717551018&KeyName=gprd-artifacts-cdn&Signature=ynzZa1NFR87JOze2xUoX7bce9yA=).

One thing I will point out however is that I had to add support for using the given IAM role ARN and WebIdentity token to the `deleteS3Bucket` functionality in the integration test suite, otherwise the call would fail as our CI pipelines do not run with any static AWS credentials. I do not know if your pipelines need this or not, but it should work as written as long as the configured IAM role for the integration tests has sufficient AWS permissions.

Please let me know if there is anything else I can do here. While I can easily help with configuring GitLab pipelines in this manner, unfortunately I do not have experience with configuring GitHub Actions or CircleCI in this manner.

denis256 commented:
```
docs/_docs/04_reference/config-blocks-and-attributes.md:1350 MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- GitLab: [Configure OpenID Co..."]

Exited with code exit status 1

```

yhakbar commented on aws_helper/config.go:
```diff
@@ -139,8 +177,13 @@ func CreateAwsSession(config *AwsSessionConfig, terragruntOptions *options.Terra
-			terragruntOptions.Logger.Debugf("Assuming role %s", terragruntOptions.IAMRoleOptions.RoleARN)
-			sess.Config.Credentials = getSTSCredentialsFromIAMRoleOptions(sess, terragruntOptions.IAMRoleOptions)
+			if terragruntOptions.IAMRoleOptions.WebIdentityToken != "" {
+				terragruntOptions.Logger.Debugf("Assuming role %s with WebIdentity token %s", terragruntOptions.IAMRoleOptions.RoleARN, terragruntOptions.IAMRoleOptions.WebIdentityToken)
```
```suggestion
				terragruntOptions.Logger.Debugf("Assuming role %s", terragruntOptions.IAMRoleOptions.RoleARN)
```
Please avoid logging the Web Identity Token

partcyborg replied:
Thanks for catching this @yhakbar. I have removed the actual token from the log message, but left the part about using a WebIdentity token in place so that it is clear from the debug logs which code path is being used.

denis256 commented:
I was wondering why the integration test passed with WebIdentityFile 
```
--- PASS: TestTerragruntAssumeRoleWebIdentityFile (13.67s)
```


I also found in internal tests that usage of WebIdentity file fails: 

```
time=2024-06-05T18:11:01Z level=error msg=Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
time=2024-06-05T[18](https://github.com/gruntwork-test/testing-terragrunt-with-web-identity/actions/runs/9389092410/job/25855946545#step:6:19):11:01Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1time=2024-06-05T18:11:01Z level=error msg=Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
time=2024-06-05T[18](https://github.com/gruntwork-test/testing-terragrunt-with-web-identity/actions/runs/9389092410/job/25855946545#step:6:19):11:01Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

Added fix in https://github.com/partcyborg/terragrunt/pull/1

partcyborg commented:
> I was wondering why the integration test passed with WebIdentityFile
> 
> ```
> --- PASS: TestTerragruntAssumeRoleWebIdentityFile (13.67s)
> ```
> 
> I also found in internal tests that usage of WebIdentity file fails:
> 
> ```
> time=2024-06-05T18:11:01Z level=error msg=Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
> 	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
> time=2024-06-05T[18](https://github.com/gruntwork-test/testing-terragrunt-with-web-identity/actions/runs/9389092410/job/25855946545#step:6:19):11:01Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1time=2024-06-05T18:11:01Z level=error msg=Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
> 	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
> time=2024-06-05T[18](https://github.com/gruntwork-test/testing-terragrunt-with-web-identity/actions/runs/9389092410/job/25855946545#step:6:19):11:01Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1
> ```
> 
> Added fix in [partcyborg#1](https://github.com/partcyborg/terragrunt/pull/1)

Interesting that it works in our environment but not yours, as when initially setting up our CI environment to run these tests I was able to execute a static rendering of `test/fixture-assume-role-web-identity/file-path` using the same input vars as the test.

Regardless, I accepted the PR and updated our private fork and the tests still pass, so the change LGTM

partcyborg commented:
> ```
> docs/_docs/04_reference/config-blocks-and-attributes.md:1350 MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- GitLab: [Configure OpenID Co..."]
> 
> Exited with code exit status 1
> ```

Added a blank line before the start of the list.

Can you point me at instructions on how to run this test? A google search for "MD032/blanks-around-lists" points me at [markdownlint](https://github.com/markdownlint/markdownlint), but running that against the docs returns a lot of unrelated issues.

yhakbar commented:
> Can you point me at instructions on how to run this test? A google search for "MD032/blanks-around-lists" points me at [markdownlint](https://github.com/markdownlint/markdownlint), but running that against the docs returns a lot of unrelated issues.

Different [markdownlint](https://github.com/DavidAnson/markdownlint).

It looks good to me!

```bash
$ markdownlint --disable 'MD013' -- docs
```

I will circle back to this tomorrow to review if Denis doesn't beat me to it. I think we're close to merging this!

yhakbar reviewed (dismissed):
Looks good to me. Will give Denis a final chance to object before merging.

GeertWille commented:
This would be an awesome addition. When can this be merged and released?

denis256 commented on aws_helper/config.go:
```diff
@@ -107,6 +113,38 @@ func CreateAwsSessionFromConfig(config *AwsSessionConfig, terragruntOptions *opt
+
+// Implements the stscreds.TokenFetcher interface
+// Supports providing a token value or the path to a token on disk
+func (f tokenFetcher) FetchToken(ctx credentials.Context) ([]byte, error) {
```
Comment should have the following format 'FetchToken ...'

partcyborg replied:
Done.

denis256 commented on aws_helper/config.go:
```diff
@@ -107,6 +113,38 @@ func CreateAwsSessionFromConfig(config *AwsSessionConfig, terragruntOptions *opt
+	}
+	token, err := os.ReadFile(string(f))
+	if err != nil {
+		return nil, err
```
`errors.WithStack(err)`

https://terragrunt.gruntwork.io/docs/community/contributing/#error-handling

partcyborg replied:
Done.

denis256 commented on aws_helper/config.go:
```diff
@@ -188,6 +235,35 @@ func AssumeIamRole(iamRoleOpts options.IAMRoleOptions) (*sts.Credentials, error)
 		sessionDurationSeconds = iamRoleOpts.AssumeRoleDuration
 	}
 
+	if iamRoleOpts.WebIdentityToken != "" {
```
Nested block can be simplified
```
if iamRoleOpts.WebIdentityToken == "" {
// return 
}
// continue flow when WebIdentityToken is set

partcyborg replied:
It isn't as simple as putting a `return` here, but I flipped the logic flow so AssumeRoleWithWebIdentity is the default case.

partcyborg commented:
@denis256 I responded to all of your review feedback. Mind taking another look?

Thanks!

levkohimins reviewed (approved):
LGTM!

dhmw commented:
This change has been helpful, thanks, but I have to point out there's a typo in one of the env var names

`TERRRAGRUNT_IAM_ASSUME_ROLE_WEB_IDENTITY_TOKEN`

You've got too many `R`s in `TERRRAGRUNT`.

yhakbar commented:
> This change has been helpful, thanks, but I have to point out there's a typo in one of the env var names
> 
> `TERRRAGRUNT_IAM_ASSUME_ROLE_WEB_IDENTITY_TOKEN`
> 
> You've got too many `R`s in `TERRRAGRUNT`.

It's an angry environment variable 😂 . Thanks for reporting. We'll look to address it.
